hal-core/src/addr.rs

```diff
@@ -0,0 +1,262 @@
+use core::{fmt, ops};
+
+pub trait Address:
+    Copy
+    + ops::Add<usize, Output = Self>
+    + ops::Sub<usize, Output = Self>
+    + ops::AddAssign<usize>
+    + ops::SubAssign<usize>
+    + PartialEq
+    + Eq
+    + PartialOrd
+    + Ord
+    + fmt::Debug
+{
+    fn as_usize(self) -> usize;
+
+    /// Aligns `self` up to `align`.
+    ///
+    /// The specified alignment must be a power of two.
+    fn align_up<A: Into<usize>>(self, align: A) -> Self;
+
+    /// Aligns `self` down to `align`.
+    ///
+    /// The specified alignment must be a power of two.
+    fn align_down<A: Into<usize>>(self, align: A) -> Self;
+
+    /// Offsets this address by `offset`.
+    ///
+    /// If the specified offset would overflow, this function saturates instead.
+    fn offset(self, offset: i32) -> Self {
+        if offset > 0 {
+            self + offset as usize
+        } else {
+            let offset = -offset;
+            self - offset as usize
+        }
+    }
+
+    /// Returns the difference between `self` and `other`.
+    fn difference(self, other: Self) -> isize {
+        if self > other {
+            -(self.as_usize() as isize - other.as_usize() as isize)
+        } else {
+            (other.as_usize() - self.as_usize()) as isize
+        }
+    }
+
+    /// Returns `true` if `self` is aligned on the specified alignment.
+    fn is_aligned<A: Into<usize>>(self, align: A) -> bool {
+        self.align_down(align) == self
+    }
+
+    fn as_ptr(&self) -> *const ();
```

Comment:
I wonder if it'd be nice to do `as_ptr<T>(&self) -> *const T`?

Out of curiosity, is there a particular reason we're using `*const T` instead of `*mut T`? Because `*mut` can implicitly downcast, I find methods returning it in unsafe code a bit nicer to work with.

Reply:
i changed it to return `*const T` in my paging branch:
https://github.com/hawkw/mycelium/blob/93188e8e171e99c3a4f991e3fdfecdefcc0ecc8a/hal-core/src/lib.rs#L81-L84

so i didn't want to touch it here. 

i'm also open to changing it to return `*mut T`; my thought process was that it was better to default to the slightly more restrictive `*const`, and if you want mutability, you have to cast to `*mut` explicitly (or we add an `as_mut_ptr` method). this may not matter.

Reply:
Most types in the stdlib which deal with raw pointers just work with `*mut T`, as by the point where you're working with raw pointers, the extra syntactic salt isn't really helpful, and can obscure the actual dangerous code. 

e.g. `NonNull` has the variance properties of `*const T`, but the `as_ptr` method always returns a `*mut T`. https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_ptr